pkg/workloadmeta/collectors/internal/remoteworkloadmeta/remoteworkloadmeta.go

```diff
@@ -0,0 +1,242 @@
+		return err
+	}
+
+	log.Info("remote workloadmeta initialized successfully")
+
+	go c.run()
+
+	return nil
+}
+
+func (c *collector) Pull(ctx context.Context) error {
+	return nil
+}
+
+func (c *collector) startWorkloadmetaStream(maxElapsed time.Duration) error {
+	expBackoff := backoff.NewExponentialBackOff()
+	expBackoff.InitialInterval = 500 * time.Millisecond
+	expBackoff.MaxInterval = 5 * time.Minute
+	expBackoff.MaxElapsedTime = maxElapsed
+
+	return backoff.Retry(func() error {
+		select {
+		case <-c.ctx.Done():
+			return &backoff.PermanentError{Err: errWorkloadmetaStreamNotStarted}
+		default:
+		}
+
+		token, err := security.FetchAuthToken()
+		if err != nil {
+			err = fmt.Errorf("unable to fetch authentication token: %w", err)
+			log.Infof("unable to establish stream, will possibly retry: %s", err)
+			return err
+		}
+
+		c.streamCtx, c.streamCancel = context.WithCancel(
+			metadata.NewOutgoingContext(
+				c.ctx,
+				metadata.MD{
+					"authorization": []string{
+						fmt.Sprintf("Bearer %s", token),
+					},
+				},
+			),
+		)
+
+		c.stream, err = c.client.WorkloadmetaStreamEntities(
+			c.streamCtx,
+			&pb.WorkloadmetaStreamRequest{
+				Filter: nil, // Subscribes to all events
+			},
+		)
+
+		if err != nil {
+			log.Infof("unable to establish stream, will possibly retry: %s", err)
+			return err
+		}
+
+		log.Info("workloadmeta stream established successfully")
+
+		return nil
+	}, expBackoff)
+}
+
+func (c *collector) run() {
+	for {
+		select {
+		case <-c.ctx.Done():
+			return
+		default:
+		}
+
+		if c.stream == nil {
+			if err := c.startWorkloadmetaStream(noTimeout); err != nil {
+				log.Warnf("error received trying to start stream: %s", err)
+				continue
+			}
+		}
+
+		var response *pb.WorkloadmetaStreamResponse
+		err := grpcutil.DoWithTimeout(func() error {
+			var err error
+			response, err = c.stream.Recv()
+			return err
+		}, streamRecvTimeout)
+		if err != nil {
+			c.streamCancel()
+
+			telemetry.RemoteStreamErrors.Inc()
+
+			// when Recv() returns an error, the stream is aborted and the
+			// contents of our store are considered out of sync. The stream must
+			// be re-established. No events are notified to the store until the
+			// connection is re-established.
+			c.stream = nil
+			c.resyncNeeded = true
+
+			log.Warnf("error received from remote workloadmeta: %s", err)
+
+			continue
+		}
+
+		err = c.processResponse(response)
+		if err != nil {
+			log.Warnf("error processing event received from remote workloadmeta: %s", err)
+			continue
+		}
+	}
+}
+
+func (c *collector) processResponse(response *pb.WorkloadmetaStreamResponse) error {
+	var collectorEvents []workloadmeta.CollectorEvent
+
+	for _, protoEvent := range response.Events {
+		workloadmetaEvent, err := utils.WorkloadmetaEventFromProtoEvent(protoEvent)
+		if err != nil {
+			return err
+		}
+
+		collectorEvent := workloadmeta.CollectorEvent{
+			Type:   workloadmetaEvent.Type,
+			Source: workloadmeta.SourceRemoteWorkloadmeta,
```

Comment:
Right. I'll check how we can keep the original source in this case.

Reply:
As we discussed off-line, workloadmeta subscribers don't get the source in the events that they receive. We'd need to change that so that the remote-workloadmeta can keep the original source.

As far as I know, filtering by source is only used in the core-agent, and not in the agents that might enable remote-workloadmeta. Also, we're only filtering by source in AD and the container lifecycle check, and it seems that filtering by source is not really adding anything. It looks like filtering by Kind should be enough. At this point, it seems that filtering by source might not be needed.

For that reason, I'm leaning towards leaving this as it is right now. We can always reconsider this when we find a use case where both filtering by source and using remote workloadmeta are required.